cnqr/_src/convex.py

```diff
@@ -84,76 +98,48 @@ def setup(self):
             bias_init=lambda *args, **kwargs: mean,
             use_bias=True,
         )
-        self.q = q
 
         # ===== Context network =====
         # (now simple dense, then will be replaced with lipschitz)
         # dense layers (with bias), init: weights -> identity, bias -> 0
         f_x = []
-        for i in range(0, self.num_hidden + 1):
-            output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
+        for i in range(0, num_hidden):
             f_x.append(
                 self.context_dense_layer(
-                    output_dim,
+                    self.dim_hidden[i],
                     name=f'f_x_dense_{i}',
                     kernel_init=self._identity_initializer,
                     bias_init=nn.initializers.constant(0.0),
                     use_bias=True,
                 )
             )
-        self.f_x = f_x
 
         # ===== Context modulators =====
         # dense layers (with bias), init: weights -> approx(0), bias -> 1
-        f_u = []
-        for i in range(0, self.num_hidden + 1):
-            output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
-            f_u.append(
-                self.context_dense_layer(
-                    output_dim,
-                    name=f'f_u_dense_{i}',
-                    kernel_init=self.init_fn(self.init_std),
-                    bias_init=nn.initializers.constant(1.0),
-                    use_bias=True,
+        f = dict()  # f_u, f_v, f_w
+        output_dims_uw = list(self.dim_hidden) + [1]
+        output_dims_v = [self.dim_data] * num_hidden + [1]
+        for module_name, output_dims in zip(['f_u', 'f_v', 'f_w'],
```

Comment:
maybe there is too much factorization now: you are using `nn.initializers.constant(1.0)` for every bias in every function, which in not compliant with the initialization scheme proposed.  
  
I believe keeping 3 for loops is better for clarity (and keeping track of everything), my suggestion was just to use `enumerate` instead of the one-liner `if i == self.num_hidden else`.